src/typechecker.erl

```diff
@@ -1586,6 +1581,79 @@ type_check_lc(Env, Expr, [Guard | Quals]) ->
     {TyL, VarBinds2, Cs2} = type_check_lc(Env#env{}, Expr, Quals),
     {TyL, union_var_binds(VarBinds1, VarBinds2), constraints:combine(Cs1, Cs2)}.
 
+%% Binary comprehension. This function is very similar to type_check_lc.
+type_check_bc(Env, Expr, []) ->
+    {Ty, _VB, Cs} = type_check_expr(Env, Expr),
+    RetTy = case normalize(Ty, Env#env.tenv) of
```

Comment:
I can confirm that. However, we may want to keep non-normalized types from `type_check_expr/2` in case of types propagated from specs, for better error messages. Possibly.

Currently I think we fail on things like `<<_/binary>> = integer_to_binary(1)` because `add_type_pat` expects normalized type and `binary()` doesn't get normalized to `<<_:_*8>>` here. I can add a known_problem for this when I get home.

Reply:
Good point! We clearly need to think more about the invariants here. It's not clear to me what the right thing actually is. I don't have the brain cycles to work on it now though. I hope to come back to it if no one else tackles it first.